homeassistant/components/zwave/cover.py

```diff
@@ -184,3 +204,105 @@ def close_cover(self, **kwargs):
     def open_cover(self, **kwargs):
         """Open the garage door."""
         self.values.primary.data = "Opened"
+
+
+class FibaroFGRM222(ZwaveRollershutter):
+    """Implementation of proprietary features for Fibaro FGR-222 / FGRM-222.
+
+    This adds support for the tile feature for the ventian blind mode.
+    For this to work, you need to set the Zwave device configuration
+    to veneian blind mode with reports type to Fibaro.
+
+    To enable this you need to configure the devices to use the venetian blind
+    mode and to enable the proprietary command class:
+    * Set "3: Reports type to Blind position reports sent"
+        to value "the main controller using Fibar Command Class"
+    * Set "10: Roller Shutter operating modes"
+        to  value "2 - Venetian Blind Mode, with positioning"
+    """
+
+    def __init__(self, hass, values, invert_buttons):
+        """Initialize the FGRM-222 with tilt mode enabled."""
+        self._value_blinds = None
+        self._value_tilt = None
+        self._has_tilt_mode = False
+        super().__init__(hass, values, invert_buttons)
+
+    @property
+    def current_cover_tilt_position(self) -> int:
+        """Get the tilt of the blinds.
+
+        Saturate values <5 and >95 so that it's easier to detect the end
+        positions in automations.
+        """
+        if not self._has_tilt_mode:
+            return None
+        if self._value_tilt.data <= 5:
+            return 0
+        elif self._value_tilt.data >= 95:
+            return 100
+        return self._value_tilt.data
+
+    def set_cover_tilt_position(self, tilt_position, **kwargs):
+        """Move the cover tilt to a specific position."""
+        if not self._has_tilt_mode:
+            _LOGGER.error("Can't set cover tilt as device is not yet set up.")
+            return
+        _LOGGER.debug("setting tilt to %d", tilt_position)
+        self._value_tilt.data = tilt_position
+
+    def open_cover_tilt(self, **kwargs):
+        """Set slats to horizontal position."""
+        self.set_cover_tilt_position(50)
```

Comment:
I just checked the values reported by the Fibaros for my venetians. Fully open, aka horizontal, is integer 99. Fully closed is 0.

When using the slider in the HA UI, setting it to 100 is not working (being ignored by the Fibaro?):
```
2019-06-08 14:29:25.882 Info, Node028, Value::Set - COMMAND_CLASS_MANUFACTURER_PROPRIETARY - Venetian blind tilt position - 1 - 1 - 100
2019-06-08 14:29:25.882 Detail, Node028, Queuing (Send) ManufacturerProprietary_SetValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x09, 0x91, 0x01, 0x0f, 0x26, 0x01, 0x01, 0x00, 0x64, 0x25, 0xc0, 0xce
2019-06-08 14:29:25.882 Detail, Node028, Queuing (Send) ManufacturerProprietary_RequestValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x08, 0x91, 0x01, 0x0f, 0x26, 0x02, 0x02, 0x00, 0x00, 0x25, 0xc1, 0xaa
2019-06-08 14:29:25.882 Detail, 
2019-06-08 14:29:25.882 Info, Node028, Sending (Send) message (Callback ID=0xc0, Expected Reply=0x04) - ManufacturerProprietary_SetValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x09, 0x91, 0x01, 0x0f, 0x26, 0x01, 0x01, 0x00, 0x64, 0x25, 0xc0, 0xce
2019-06-08 14:29:25.893 Detail, Node028,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-06-08 14:29:25.893 Detail, Node028,   ZW_SEND_DATA delivered to Z-Wave stack
2019-06-08 14:29:25.903 Detail, Node028,   Received: 0x01, 0x07, 0x00, 0x13, 0x05, 0x00, 0x00, 0x01, 0xef
2019-06-08 14:29:25.903 Detail, Node028,   ZW_SEND_DATA Request with callback ID 0x05 received (expected 0x05)
2019-06-08 14:29:25.903 Info, Node028, Request RTT 21 Average Request RTT 26
2019-06-08 14:29:35.889 Error, Node028, ERROR: Dropping command, expected response not received after 1 attempt(s)
2019-06-08 14:29:35.889 Detail, Node028, Removing current message
2019-06-08 14:29:35.889 Detail, Node028, Notification: Notification - TimeOut
2019-06-08 14:29:35.891 Detail, 
2019-06-08 14:29:35.891 Info, Node028, Sending (Send) message (Callback ID=0xc1, Expected Reply=0x04) - ManufacturerProprietary_RequestValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x08, 0x91, 0x01, 0x0f, 0x26, 0x02, 0x02, 0x00, 0x00, 0x25, 0xc1, 0xaa
2019-06-08 14:29:35.915 Detail, Node028,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-06-08 14:29:35.915 Detail, Node028,   ZW_SEND_DATA delivered to Z-Wave stack
2019-06-08 14:29:35.933 Detail, Node028,   Received: 0x01, 0x07, 0x00, 0x13, 0xc1, 0x00, 0x00, 0x02, 0x28
2019-06-08 14:29:35.933 Detail, Node028,   ZW_SEND_DATA Request with callback ID 0xc1 received (expected 0xc1)
2019-06-08 14:29:35.933 Info, Node028, Request RTT 41 Average Request RTT 33
2019-06-08 14:29:35.933 Detail,   Expected callbackId was received
2019-06-08 14:29:35.943 Detail, Node028,   Received: 0x01, 0x0e, 0x00, 0x04, 0x00, 0x1c, 0x08, 0x91, 0x01, 0x0f, 0x26, 0x03, 0x03, 0x06, 0x63, 0x3d
2019-06-08 14:29:35.943 Detail, 
2019-06-08 14:29:35.943 Info, Node028, Response RTT 52 Average Response RTT 515
2019-06-08 14:29:35.944 Info, Node028, Received Fibaro proprietary blind/slat position for node 28: Blinds: 6 Slats: 99
2019-06-08 14:29:35.944 Detail, Node028, Refreshed Value: old value=6, new value=6, type=byte
2019-06-08 14:29:35.944 Detail, Node028, Changes to this value are not verified
2019-06-08 14:29:35.944 Detail, Node028, Refreshed Value: old value=99, new value=99, type=byte
2019-06-08 14:29:35.944 Detail, Node028, Changes to this value are not verified
2019-06-08 14:29:35.944 Detail, Node028,   Expected reply and command class was received
2019-06-08 14:29:35.944 Detail, Node028,   Message transaction complete
2019-06-08 14:29:35.944 Detail, 
2019-06-08 14:29:35.944 Detail, Node028, Removing current message
2019-06-08 14:29:35.944 Detail, Node028, Notification: ValueChanged
2019-06-08 14:29:35.953 Detail, Node028, Notification: ValueChanged
```

Using the slider to set it to 99 works flawlessly, though:
```
2019-06-08 14:33:58.016 Info, Node028, Value::Set - COMMAND_CLASS_MANUFACTURER_PROPRIETARY - Venetian blind tilt position - 1 - 1 - 99
2019-06-08 14:33:58.016 Detail, Node028, Queuing (Send) ManufacturerProprietary_SetValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x09, 0x91, 0x01, 0x0f, 0x26, 0x01, 0x01, 0x00, 0x63, 0x25, 0xc8, 0xc1
2019-06-08 14:33:58.016 Detail, Node028, Queuing (Send) ManufacturerProprietary_RequestValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x08, 0x91, 0x01, 0x0f, 0x26, 0x02, 0x02, 0x00, 0x00, 0x25, 0xc9, 0xa2
2019-06-08 14:33:58.016 Detail, 
2019-06-08 14:33:58.016 Info, Node028, Sending (Send) message (Callback ID=0xc8, Expected Reply=0x04) - ManufacturerProprietary_SetValue (Node=28): 0x01, 0x0f, 0x00, 0x13, 0x1c, 0x09, 0x91, 0x01, 0x0f, 0x26, 0x01, 0x01, 0x00, 0x63, 0x25, 0xc8, 0xc1
2019-06-08 14:33:58.029 Detail, Node028,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-06-08 14:33:58.029 Detail, Node028,   ZW_SEND_DATA delivered to Z-Wave stack
2019-06-08 14:33:58.039 Detail, Node028,   Received: 0x01, 0x07, 0x00, 0x13, 0x05, 0x00, 0x00, 0x02, 0xec
2019-06-08 14:33:58.040 Detail, Node028,   ZW_SEND_DATA Request with callback ID 0x05 received (expected 0x05)
2019-06-08 14:33:58.040 Info, Node028, Request RTT 23 Average Request RTT 25
2019-06-08 14:33:59.202 Detail, Node028,   Received: 0x01, 0x0c, 0x00, 0x04, 0x00, 0x1c, 0x06, 0x31, 0x05, 0x04, 0x22, 0x04, 0x3a, 0xc1
2019-06-08 14:33:59.202 Detail, 
2019-06-08 14:33:59.203 Info, Node028, Response RTT 1186 Average Response RTT 1177
2019-06-08 14:33:59.203 Info, Node028, Received SensorMultiLevel report from node 28, instance 1, Power: value=108.2W
2019-06-08 14:33:59.203 Detail, Node028, Refreshed Value: old value=0.0, new value=108.2, type=decimal
2019-06-08 14:33:59.203 Detail, Node028, Changes to this value are not verified
2019-06-08 14:33:59.203 Detail, Node028, Notification: ValueChanged
2019-06-08 14:34:00.286 Detail, Node028,   Received: 0x01, 0x0e, 0x00, 0x04, 0x00, 0x1c, 0x08, 0x91, 0x01, 0x0f, 0x26, 0x03, 0x03, 0x09, 0x63, 0x32
2019-06-08 14:34:00.286 Detail, 
2019-06-08 14:34:00.286 Info, Node028, Response RTT 2269 Average Response RTT 1723
2019-06-08 14:34:00.286 Info, Node028, Received Fibaro proprietary blind/slat position for node 28: Blinds: 9 Slats: 99
2019-06-08 14:34:00.286 Detail, Node028, Refreshed Value: old value=8, new value=9, type=byte
2019-06-08 14:34:00.286 Detail, Node028, Changes to this value are not verified
2019-06-08 14:34:00.286 Detail, Node028, Refreshed Value: old value=0, new value=99, type=byte
2019-06-08 14:34:00.287 Detail, Node028, Changes to this value are not verified
2019-06-08 14:34:00.287 Detail, Node028, Notification: ValueChanged
2019-06-08 14:34:00.293 Detail, Node028, Notification: ValueChanged
2019-06-08 14:34:00.304 Detail, Node028,   Received: 0x01, 0x0c, 0x00, 0x04, 0x00, 0x1c, 0x06, 0x31, 0x05, 0x04, 0x22, 0x00, 0x00, 0xff
2019-06-08 14:34:00.304 Detail, 
2019-06-08 14:34:00.304 Info, Node028, Response RTT 2287 Average Response RTT 2005
2019-06-08 14:34:00.304 Info, Node028, Received SensorMultiLevel report from node 28, instance 1, Power: value=0.0W
2019-06-08 14:34:00.304 Detail, Node028, Refreshed Value: old value=108.2, new value=0.0, type=decimal
2019-06-08 14:34:00.304 Detail, Node028, Changes to this value are not verified
2019-06-08 14:34:00.304 Detail, Node028, Notification: ValueChanged
```

@ChristianKuehnel Do you observe a similar behavior? If so, should we put a ceiling function there that does `100->99`?
Sould this be  `self.set_cover_tilt_position(99)` instead of 50?

Reply:
By the way, the Fibaro expecting a range of `[0-99]` makes sense, since this range has 100 values in total. The HA UI slider allows to set `[0-100]`, which is 101 values.

Reply:
@andre-richter 

Thx for the fast review!

I can confirm setting the tilt to 100 doesn't do anything. So I added a limit of `[0-99]`. I would still show 100 as maximum value, as this is the same for the `cover_position`. And "set slats to 100%" feels more natural than "set slats to 99%".

Concerning the semantics of the values, this might be more difficult. For my setup I have:
* 0 = outside edge facing downwards, I would call this **closed**
* 50 = slats are horizontal , this what I used to call **open**
* 99 = outside edge facing upwards, I'm not using this.

I understand for your setup:
* 0 = same as for me
* 50 = angle of 45°, outside edge facing down
* 100 = horizontal
* So what is the value when the *outside edge is facing upwards*?

Config parameter 12: "In Venetian Blind mode (parameter 10 set to 2) the parameter determines time of full turn of the lamellas." For my setup the default value of 1.5 sec works quite nicely. With this configuration I can set the full range of 0-99 (see above) and 50 the slats are horizontal.

Reply:
Your description of my setup is correct.

> So what is the value when the outside edge is facing upwards?

Apparently the motors of my blinds only support turning the slats between closed and horizontal. I cannot go any higher, like letting the outside edge point towards the sky. Until now, I didn't know this was possible :see_no_evil:.
For config parameter 12, a value of `120` fits my setup. Anything greater than `120` will just start to pull the blinds up alltogether, but doesn't add any more rotation.

Not sure what an elegant solution to this problem is. Adding a `configuration variable` like [invert_openclose_buttons](https://www.home-assistant.io/docs/z-wave/installation/#invert_openclose_buttons) that allows to override the default value that `open_cover_tilt()` is setting? Would be a bit annoying if you have many shutter controllers, but I fail to come up with a better idea right now.

Reply:
What happens when you on your setup parameter 12 is set to 0.75 seconds, and you set the range to 99? Where do your slats end up?

Reply:
First off: The `open_cover_tilt` function is just a convenience feature. If you don't like it, you can always set the tilt angle manually. I guess in most cases people will call this from some script or automation anyway. So if it does not work perfectly, it should not be a big problem.

Setting parameter 12 to the 1/2 rotation time with my blinds works when moving from "edge down" to "horizontal", but I can't move from "edge up" to "edge down" any more.

I see two ways of going forward:
- Flipping a coin and hard cording "open" to 50 or 100.
- Adding an option to let the user configure what "open" means for them, so not only a binary option, but allow for `[0-99]`.

What do you think?